scio-core/src/main/scala/com/spotify/scio/util/ArtisanJoin.scala

```diff
@@ -0,0 +1,94 @@
+import org.apache.beam.sdk.transforms.{DoFn, ParDo}
+import org.apache.beam.sdk.transforms.join.{CoGbkResult, CoGroupByKey, KeyedPCollectionTuple}
+import org.apache.beam.sdk.values.{KV, TupleTag}
+
+import scala.reflect.ClassTag
+
+private[scio] object ArtisanJoin {
+
+  private def cogroup[KEY: ClassTag, A: ClassTag, B: ClassTag, A1, B1]
+  (name: String, a: SCollection[(KEY, A)], b: SCollection[(KEY, B)])
+  (leftFn: JIterable[A] => JIterable[A1], rightFn: JIterable[B] => JIterable[B1])
+  : SCollection[(KEY, (A1, B1))] = {
+    val (tagA, tagB) = (new TupleTag[A](), new TupleTag[B]())
+    val keyed = KeyedPCollectionTuple
+      .of(tagA, a.toKV.internal)
+      .and(tagB, b.toKV.internal)
+      .apply(name, CoGroupByKey.create())
+
+    type DF = DoFn[KV[KEY, CoGbkResult], (KEY, (A1, B1))]
+    a.context.wrap(keyed).withName(name)
+      .applyTransform(ParDo.of(new DF {
+        @ProcessElement
+        private[util] def processElement(c: DF#ProcessContext): Unit = {
+          val kv = c.element()
+          val key = kv.getKey
+          val result = kv.getValue
+          val as = leftFn(result.getAll(tagA))
+          val bs = rightFn(result.getAll(tagB))
+          val ai = as.iterator()
+          while (ai.hasNext) {
+            val a = ai.next()
+            val bi = bs.iterator()
+            while (bi.hasNext) {
+              c.output((key, (a, bi.next())))
+            }
+          }
+        }
+      }))
+  }
+
+  def apply[KEY: ClassTag, A: ClassTag, B: ClassTag](name: String,
+                                                     a: SCollection[(KEY, A)],
+                                                     b: SCollection[(KEY, B)])
+  : SCollection[(KEY, (A, B))] = cogroup(name, a, b)(identity, identity)
+
+  def left[KEY: ClassTag, A: ClassTag, B: ClassTag](name: String,
+                                                    a: SCollection[(KEY, A)],
+                                                    b: SCollection[(KEY, B)])
+  : SCollection[(KEY, (A, Option[B]))] = cogroup(name, a, b)(identity, toOptions)
+
+  def right[KEY: ClassTag, A: ClassTag, B: ClassTag](name: String,
+                                                     a: SCollection[(KEY, A)],
+                                                     b: SCollection[(KEY, B)])
+  : SCollection[(KEY, (Option[A], B))] = cogroup(name, a, b)(toOptions, identity)
+
+  def outer[KEY: ClassTag, A: ClassTag, B: ClassTag](name: String,
+                                                     a: SCollection[(KEY, A)],
+                                                     b: SCollection[(KEY, B)])
+  : SCollection[(KEY, (Option[A], Option[B]))] = cogroup(name, a, b)(toOptions, toOptions)
+
+  private def toOptions[A](xs: JIterable[A]): JIterable[Option[A]] =
+    if (xs.iterator().hasNext) {
+      Iterables.transform(xs, new com.google.common.base.Function[A, Option[A]] {
+        override def apply(input: A) = Option(input)
```

Comment:
Should this be `Some(input)`, instead of `Option` specifically if the RHS contains valid `null` that might infer `None` for a valid `null` RHS value.

Reply:
We should never have null input here since there's no null coder. So `Some` or `Option` doesn't make a huge difference.

Reply:
One less `if` then :)

Reply:
Can't really get rid of the `if` here. Empty list => `List(None)`, non-empty list => `xs.map(Some(_))`. I'll change `Option(input)` to `Some(input)` but that's only for readability.

Reply:
I mean inside the `Option`, there is `if` null check that goes to `Some` or `None`, but since this is always `Some`, should make sense to just use `Some`, but yes - readability.

Reply:
🤣 the invisible `if`